File: example/src/main/java/com/github/vivchar/example/pages/simple/YourDataProvider.java
package com.github.vivchar.example.pages.simple;

import androidx.recyclerview.widget.ListUpdateCallback;

import com.github.vivchar.example.pages.simple.DiffUtilFragment.DiffViewModel;
import com.github.vivchar.example.pages.simple.PayloadFragment.PayloadViewModel;
import com.github.vivchar.rendererrecyclerviewadapter.DefaultCompositeViewModel;
import com.github.vivchar.rendererrecyclerviewadapter.ViewModel;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

import static com.github.vivchar.example.pages.simple.ViewRendererFragment.RectViewModel;
import static com.github.vivchar.example.pages.simple.ViewStateFragment.StateViewModel;

/**
 * Created by Vivchar Vitaly on 28.12.17.
 */

public class YourDataProvider {

	final ArrayList<PayloadViewModel> mPayloadItems = new ArrayList<>();

	private ArrayList<LoadMoreFragment.SimpleViewModel> mLoadMoreItems = new ArrayList<>();
	private ThreadPoolExecutor mExecutor = new ThreadPoolExecutor(2, 2, 2, TimeUnit.SECONDS, new LinkedBlockingQueue<>());
	private ArrayList<ViewModel> mDiffItems = new ArrayList<>();

	public List<ViewModel> getDiffItems() {
		for (int i = 0; i < 50; i++) {
			mDiffItems.add(new DiffViewModel(i, String.valueOf(i)));
		}
		return mDiffItems;
	}

	public List<ViewModel> getUpdatedDiffItems(final DiffViewModel model) {
		final int clickedIndex = mDiffItems.indexOf(model);
		final ViewModel clickedModel = mDiffItems.remove(clickedIndex);
		final ViewModel remove = mDiffItems.remove(0);
		Collections.shuffle(mDiffItems);

		/* https://stackoverflow.com/a/43461324/4894238
		 * use RendererRecyclerViewAdapter.setUpdateCallback(ListUpdateCallback) if you want
		 * */
		mDiffItems.add(0, remove);
		mDiffItems.add(clickedIndex, clickedModel);
		return mDiffItems;
	}

	public List<ViewModel> getSquareItems() {
		final ArrayList<ViewModel> items = new ArrayList<>();
		for (int i = 0; i < 50; i++) {
			items.add(new RectViewModel(i, String.valueOf(i)));
		}
		return items;
	}

	public List<ViewModel> getCompositeSimpleItems() {
		final ArrayList<ViewModel> items = new ArrayList<>();
		for (int i = 0; i < 50; i++) {
			items.add(new DefaultCompositeViewModel(getDiffItems()));
		}
		return items;
	}

	public List<ViewModel> getStateItems() {
		final ArrayList<ViewModel> items = new ArrayList<>();
		for (int i = 0; i < 50; i++) {
			items.add(new StateViewModel(i, getDiffItems()));
		}
		return items;
	}

	public List<PayloadViewModel> getPayloadItems() {
		if (mPayloadItems.isEmpty()) {
			for (int i = 0; i < 50; i++) {
				mPayloadItems.add(new PayloadViewModel(i, String.valueOf(i), "model ID: " + i));
			}
		}
		return mPayloadItems;
	}

	public List<PayloadViewModel> getChangedPayloadItems(final PayloadViewModel model) {
		/* Just for example */
		final ArrayList<PayloadViewModel> newList = new ArrayList<>();

		for (final PayloadViewModel viewModel : mPayloadItems) {
			if (viewModel.getID() == model.getID()) {
				newList.add(new PayloadViewModel(
						model.getID(),
						String.valueOf(Integer.valueOf(model.getText()) + 1),
						"model ID: " + model.getID()
				));
			} else {
				newList.add(viewModel);
			}
		}

		mPayloadItems.clear();
		mPayloadItems.addAll(newList);

		return mPayloadItems;
	}

	public List<? extends ViewModel> getLoadMoreItems() {
		final int size = mLoadMoreItems.size();
		for (int i = size; i < size + 30; i++) {
			mLoadMoreItems.add(new LoadMoreFragment.SimpleViewModel(String.valueOf(i)));
		}
		return mLoadMoreItems;
	}

	public void getLoadMoreItems(final Listener listener) {
		mExecutor.execute(() -> {
			try {
				Thread.sleep(2000);
			} catch (InterruptedException e) {
				e.printStackTrace();
			}

			final int size = mLoadMoreItems.size();
			for (int i = size; i < size + 30; i++) {
				mLoadMoreItems.add(new LoadMoreFragment.SimpleViewModel(String.valueOf(i)));
			}
			listener.onChanged(new ArrayList<>(mLoadMoreItems));
		});
	}

	public interface Listener {

		void onChanged(List<? extends ViewModel> list);
	}
}

File: example/src/main/java/com/github/vivchar/example/pages/simple/YourDataProvider.kt
package com.github.vivchar.example.pages.simple

import com.github.vivchar.example.pages.simple.DiffUtilFragment.DiffViewModel
import com.github.vivchar.example.pages.simple.InputsFragment.InputViewModel
import com.github.vivchar.example.pages.simple.LoadMoreFragment.SimpleViewModel
import com.github.vivchar.example.pages.simple.PayloadFragment.PayloadViewModel
import com.github.vivchar.example.pages.simple.ViewRendererFragment.RectViewModel
import com.github.vivchar.example.pages.simple.ViewStateFragment.StateViewModel
import com.github.vivchar.rendererrecyclerviewadapter.DefaultCompositeViewModel
import com.github.vivchar.rendererrecyclerviewadapter.ViewModel
import java.util.*
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit

/**
 * Created by Vivchar Vitaly on 28.12.17.
 */
class YourDataProvider {
	private val executor = ThreadPoolExecutor(2, 2, 2, TimeUnit.SECONDS, LinkedBlockingQueue())

	val diffItems = mutableListOf<ViewModel>().apply { for (i in 0..49) add(DiffViewModel(i, i.toString())) }
	val squareItems = mutableListOf<ViewModel>().apply { for (i in 0..49) add(RectViewModel(i, i.toString())) }
	val compositeSimpleItems = mutableListOf<ViewModel>().apply { for (i in 0..49) add(DefaultCompositeViewModel(diffItems)) }
	val inputsModels = mutableListOf<ViewModel>().apply { for (i in 0..49) add(InputViewModel(i, String.format("Value %d", i))) }
	val stateItems = mutableListOf<ViewModel>().apply { for (i in 0..49) add(StateViewModel(i, diffItems)) }
	val payloadItems = mutableListOf<PayloadViewModel>().apply { for (i in 0..49) add(PayloadViewModel(i, i.toString(), "model ID: $i")) }
	val loadMoreItems = mutableListOf<ViewModel>().apply { for (i in 0..30) add(SimpleViewModel(i.toString())) }

	fun getUpdatedDiffItems(model: DiffViewModel?): List<ViewModel?> {
		val clickedIndex = diffItems.indexOf(model)
		val clickedModel = diffItems.removeAt(clickedIndex)
		val remove = diffItems.removeAt(0)
		diffItems.shuffle()

		/* https://stackoverflow.com/a/43461324/4894238
		 use RendererRecyclerViewAdapter.setUpdateCallback(ListUpdateCallback) if you want */
		diffItems.add(0, remove)
		diffItems.add(clickedIndex, clickedModel)
		return diffItems
	}

	fun getChangedPayloadItems(model: PayloadViewModel): List<PayloadViewModel> {
		/* Just for example */
		val newList = ArrayList<PayloadViewModel>()
		for (viewModel in payloadItems) {
			if (viewModel.id == model.id) {
				newList.add(
					PayloadViewModel(
						model.id, (Integer.valueOf(model.text) + 1).toString(),
						"model ID: " + model.id
					)
				)
			} else {
				newList.add(viewModel)
			}
		}
		payloadItems.clear()
		payloadItems.addAll(newList)
		return payloadItems
	}

	fun getLoadMoreItems(listener: (List<ViewModel>) -> Unit) {
		executor.execute {
			try {
				Thread.sleep(2000)
			} catch (e: InterruptedException) {
				e.printStackTrace()
			}
			val size = loadMoreItems.size
			for (i in size until size + 30) {
				loadMoreItems.add(SimpleViewModel(i.toString()))
			}
			listener.invoke(ArrayList(loadMoreItems))
		}
	}
}